KAFKA-8470: State change logs should not be in TRACE level

stanislavkozlovski opened the pull request:
This patch revises the log level and content of many of the state change logger logs and changes its default log level to INFO.

stanislavkozlovski committed:
KAFKA-8470-state-change-logger

Select logs should stay in trace

Address comments

Add state change logger to Partition#makeLeader()
Remove per-partition logging in channelManager#sendUpdateMetadataRequest
Remove per-partition logging in MetadataCache#updateMetadata
Reduce per-partition logging in ReplicaManager

Change logging in Partition#makeLeader and Partition#makeFollower

ControllerChannelManager - Add TRACE logging in sendStopReplicaRequests and change some LAIR sending  logging to TRACE

Clear unnecessary logging in ReplicaManager and switch to TRACE

We reduce the number of for loops in become-leader/become-follower LAIR handling, log in trace (because Partition#makeFollower/makeLeader() already log per partition) and add some introductory single info-level logs

Remove unnecessary log in stopReplicas()

Address PR comments

Address PR comments

Log number of become-leader/become-follower partition states

Do not re-instantiate StateChangeLogger when sending lair/updateMetadata/stopReplica requests to brokers

Add per-broker info logs on stop replica requests

Log in trace for NonExistentPartition state switch

Log OfflineReplica state switch per partition in INFO

Log sendUpdateMetadataRequest in INFO

Log OfflinePartition in TRACE for PartitionStateMachine

Return per-partition TRACE logging in metadata update

Revise ReplicaManager log statement placement, make LeaderAndIsr ignore log INFO and reword remove replica log

stanislavkozlovski commented:
I tried to be selective in the logs we keep in trace in order to avoid a regression in log volume. This commit (https://github.com/apache/kafka/pull/8320/commits/fcbe35cc34341554e6846e96ef31495a6377cf0d) contains the logs I think we should leave in TRACE (+ [some expensive logs in the UpdateMetadataRequest path](https://github.com/apache/kafka/pull/8320/files#diff-bfeebf48d90e86c1ffb850fbbe019dd6R310)), all others were changed to INFO.

This PR is a continuation of https://github.com/apache/kafka/pull/6878

ijuma commented on core/src/main/scala/kafka/controller/ControllerChannelManager.scala:
```diff
@@ -266,9 +266,10 @@ class RequestSendThread(val controllerId: Int,
+        if (stateChangeLogger.isTraceEnabled)
+          stateChangeLogger.withControllerEpoch(controllerContext.epoch).trace(s"Received response " +
+            s"${response.toString(requestHeader.apiVersion)} for request $api with correlation id " +
+            s"${requestHeader.correlationId} sent to broker $brokerNode")
```
Why is this trace?

junrao replied:
The main goal is to avoid partition level logging for UpdateMetadata since it includes all partitions in the cluster. So we can probably log in info, but for UpdataMetadata, only log partition count.

stanislavkozlovski replied:
There was an earlier comment on the PR suggesting we were more selective with what we log at INFO. I had considered reverting it from INFO here (https://github.com/apache/kafka/pull/6878#issuecomment-501214135) and received no comments against it, so I thought it was OK.

This isn't a per-partition log though, so I agree we don't need to keep it at trace

ijuma commented on core/src/main/scala/kafka/controller/ControllerChannelManager.scala:
```diff
@@ -447,13 +448,11 @@ abstract class AbstractControllerBrokerRequestBatch(config: KafkaConfig,
+          val typeOfRequest =
+            if (broker == state.leader) "become-leader"
+            else "become-follower"
+          stateChangeLog.info(s"Sending $typeOfRequest LeaderAndIsr request $state to broker $broker for partition $topicPartition")
```
Shouldn't this one be `debug` at least?

stanislavkozlovski replied:
I'm fine with reducing INFO level logs. My preference would be to group these and then log an info at the end `sent $typeOfRequest for partitions X`. WDYT we have both the debug and the aggregated log?

junrao replied:
We can probably keep this as trace.

ijuma commented on core/src/main/scala/kafka/controller/ControllerChannelManager.scala:
```diff
@@ -473,10 +472,11 @@ abstract class AbstractControllerBrokerRequestBatch(config: KafkaConfig,
+      updateMetadataRequestPartitionInfoMap.foreach { case (tp, partitionState) =>
+        stateChangeLog.trace(s"Sending UpdateMetadata request $partitionState to brokers $updateMetadataRequestBrokerSet " +
+          s"for partition $tp")
+      }
```
Why is this `trace`?

junrao replied:
We probably can just do an info level logging but with just the partition count.

ijuma commented on core/src/main/scala/kafka/server/MetadataCache.scala:
```diff
@@ -302,22 +302,33 @@ class MetadataCache(brokerId: Int) extends Logging {
             removePartitionInfo(partitionStates, tp.topic, tp.partition)
-            stateChangeLogger.trace(s"Deleted partition $tp from metadata cache in response to UpdateMetadata " +
-              s"request sent by controller $controllerId epoch $controllerEpoch with correlation id $correlationId")
+            if (stateChangeLogger.isTraceEnabled)
```
Instead of calling this once per partition, I'd do it once before the loop.

stanislavkozlovski replied:
Come to think of it, I agree. It's better to not let users shoot themselves in the foot if they enable TRACE state change logging

stanislavkozlovski replied:
What did you mean before the loop? Get the number of partitions who are scheduled for deleting and print we're about to delete N of those?

ijuma replied:
I meant the isTraceEnabled call since this require a bit of work in the logger to figure out if it's enabled.

stanislavkozlovski replied:
I see. Do you agree with my suggestion of not having the logs at all, even in trace?

ijuma commented on core/src/main/scala/kafka/server/MetadataCache.scala:
```diff
@@ -302,22 +302,33 @@ class MetadataCache(brokerId: Int) extends Logging {
+        stateChangeLogger.info(s"Deleted ${deletedPartitions.size} partitions from metadata cache in response to UpdateMetadata " +
+          s"request sent by controller $controllerId epoch $controllerEpoch with correlation id $correlationId (partitions: $deletedPartitions)")
+        stateChangeLogger.info(s"Cached leader info for ${cachedPartitions.size} partitions in response to UpdateMetadata " +
+          s"request sent by controller $controllerId epoch $controllerEpoch with correlation id $correlationId (partitions: $cachedPartitions)")
```
Maybe we don't include the actual partitions in these two logs? Just a count or something?

junrao replied:
Since UpdateMetadata includes all partitions for the cluster, at the info level, we could probably just log a count instead of the actual partition list.

ijuma commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -386,8 +389,10 @@ class ReplicaManager(val config: KafkaConfig,
               stateChangeLogger.error(s"Ignoring stop replica (delete=${stopReplicaRequest.deletePartitions}) for " +
                 s"partition $topicPartition due to storage exception", e)
               responseMap.put(topicPartition, Errors.KAFKA_STORAGE_ERROR)
+              stopReplicaCount -= 1
```
Why not increment instead of decrement?

stanislavkozlovski replied:
Oh, because I already had the variable present due to the log where we print out how many partitions we're about to stop:
```
        stateChangeLogger.info(s"Handling stop replica (delete=${stopReplicaRequest.deletePartitions}) for ${partitions.size} partitions (${partitions.mkString(", ")})")
```

stanislavkozlovski replied:
Is it weird to decrement?

ijuma replied:
Yeah, it's a bit less conventional.

ijuma reviewed (commented):
I left a few comments, but I'd like to hear from @junrao as well before making changes.

junrao commented on core/src/main/scala/kafka/server/MetadataCache.scala:
```diff
@@ -302,22 +302,33 @@ class MetadataCache(brokerId: Int) extends Logging {
+        val controllerEpoch = updateMetadataRequest.controllerEpoch
+        val cachedPartitions = new mutable.ArrayBuffer[TopicPartition]
+        updateMetadataRequest.partitionStates.asScala.foreach { state =>
+          // per-partition logging here proves to be very expensive due to holding up the partition lock and the request thread
```
UpdateMetadata doesn't hold partition level lock. It just holds the ReplicaStateLock. We can probably say that it's expensive since it includes all partitions in the cluster.

ijuma replied:
@junrao It also holds the `partitionMetadataLock`, so if you have multiple `UpdateMetadataRequest` calls, they each wait behind each other.

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -377,7 +377,10 @@ class ReplicaManager(val config: KafkaConfig,
         replicaAlterLogDirsManager.removeFetcherForPartitions(partitions)
-        for (topicPartition <- partitions){
+
+        var stopReplicaCount = partitions.size
```
stopReplicaCount => SuccessfulStopReplicaCount ?

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -377,7 +377,10 @@ class ReplicaManager(val config: KafkaConfig,
-        for (topicPartition <- partitions){
+
+        var stopReplicaCount = partitions.size
+        stateChangeLogger.info(s"Handling stop replica (delete=${stopReplicaRequest.deletePartitions}) for ${stopReplicaCount} partitions (${partitions.mkString(", ")})")
```
In other places, we log a partition in a separate line. So, we should be consistent.

stanislavkozlovski replied:
What do you think this should look like? The same `Handling stop replica...` message iterated N times for each partition and logged as one multi-line string?
Is it better to just not print the partitions?

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1468,7 +1471,7 @@ class ReplicaManager(val config: KafkaConfig,
       replicaFetcherManager.removeFetcherForPartitions(partitionsToMakeFollower.map(_.topicPartition))
       partitionsToMakeFollower.foreach { partition =>
-        stateChangeLogger.trace(s"Stopped fetchers as part of become-follower request from controller $controllerId " +
+        stateChangeLogger.info(s"Stopped fetchers as part of become-follower request from controller $controllerId " +
```
Here, we could just do an info level logging with the partition count, instead of actual partition list.

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1478,14 +1481,14 @@ class ReplicaManager(val config: KafkaConfig,
 
       partitionsToMakeFollower.foreach { partition =>
-        stateChangeLogger.trace(s"Truncated logs and checkpointed recovery boundaries for partition " +
+        stateChangeLogger.info(s"Truncated logs and checkpointed recovery boundaries for partition " +
```
This seems to be useless now since we moved the truncation logic inside replica fetcher. So, we can just remove this logging.

stanislavkozlovski replied:
Nice, I didn't know that.

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1353,7 +1356,7 @@ class ReplicaManager(val config: KafkaConfig,
           if (partition.makeLeader(controllerId, partitionState, correlationId, highWatermarkCheckpoints)) {
             partitionsToMakeLeaders += partition
-            stateChangeLogger.trace(s"Stopped fetchers as part of become-leader request from " +
+            stateChangeLogger.info(s"Stopped fetchers as part of become-leader request from " +
```
This logging seems mis-placed. We probably just want to move this to after replicaFetcherManager.removeFetcherForPartitions() with just the partition count. We already have info logging inside Partition.makeLeader() and we can change that logging to stateChangeLog.

junrao reviewed (commented):
@stanislavkozlovski : Thanks for the PR. The main goal of this PR is probably to remove the partition level logging for UpdateMetadata since it includes all partitions in the cluster. Left a few comments below.

junrao commented on core/src/main/scala/kafka/server/MetadataCache.scala:
```diff
@@ -302,22 +302,27 @@ class MetadataCache(brokerId: Int) extends Logging {
+        val controllerEpoch = updateMetadataRequest.controllerEpoch
+        val cachedPartitions = new mutable.ArrayBuffer[TopicPartition]
+        updateMetadataRequest.partitionStates.asScala.foreach { state =>
+          // per-partition logging here proves to be expensive because it includes all partitions in the cluster
```
This comment is no longer needed.

junrao commented on core/src/main/scala/kafka/server/MetadataCache.scala:
```diff
@@ -302,22 +302,27 @@ class MetadataCache(brokerId: Int) extends Logging {
+
+        val controllerId = updateMetadataRequest.controllerId
+        val controllerEpoch = updateMetadataRequest.controllerEpoch
+        val cachedPartitions = new mutable.ArrayBuffer[TopicPartition]
```
We just need the count. So cachedPartitions could be an int.

stanislavkozlovski replied:
Ah sorry, left over from the past change

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1334,12 +1337,6 @@ class ReplicaManager(val config: KafkaConfig,
                           correlationId: Int,
                           responseMap: mutable.Map[TopicPartition, Errors],
                           highWatermarkCheckpoints: OffsetCheckpoints): Set[Partition] = {
-    partitionStates.keys.foreach { partition =>
```
Let's keep this as TRACE logging here.

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1348,12 +1345,15 @@ class ReplicaManager(val config: KafkaConfig,
     try {
       // First stop fetchers for all the partitions
       replicaFetcherManager.removeFetcherForPartitions(partitionStates.keySet.map(_.topicPartition))
+      stateChangeLogger.info(s"Handling LeaderAndIsr request correlationId $correlationId from " +
```
Change the logging to "Stopped fetchers as part of become-leader request from ...".

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1348,12 +1345,15 @@ class ReplicaManager(val config: KafkaConfig,
           if (partition.makeLeader(controllerId, partitionState, correlationId, highWatermarkCheckpoints)) {
             partitionsToMakeLeaders += partition
-            stateChangeLogger.trace(s"Stopped fetchers as part of become-leader request from " +
+            stateChangeLogger.info(s"Stopped fetchers as part of become-leader request from " +
```
We can remove this logging since partition.makeLeader() logs to stateChangeLog now.

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1384,7 +1384,7 @@ class ReplicaManager(val config: KafkaConfig,
 
     partitionStates.keys.foreach { partition =>
-      stateChangeLogger.trace(s"Completed LeaderAndIsr request correlationId $correlationId from controller $controllerId " +
+      stateChangeLogger.info(s"Completed LeaderAndIsr request correlationId $correlationId from controller $controllerId " +
```
Let's keep this as TRACE logging here.

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1416,7 +1416,7 @@ class ReplicaManager(val config: KafkaConfig,
                             highWatermarkCheckpoints: OffsetCheckpoints) : Set[Partition] = {
     partitionStates.foreach { case (partition, partitionState) =>
-      stateChangeLogger.trace(s"Handling LeaderAndIsr request correlationId $correlationId from controller $controllerId " +
+      stateChangeLogger.info(s"Handling LeaderAndIsr request correlationId $correlationId from controller $controllerId " +
```
Let's keep this as TRACE logging here. But add an info level state change log in Partition.makeFollower().

stanislavkozlovski replied:
I'll also add a single INFO log to mention the # of partitions we're about to handle. This should make it easy to follow the sequence of operations when Partition#makeFollower starts logging

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1467,25 +1467,16 @@ class ReplicaManager(val config: KafkaConfig,
       if (isShuttingDown.get()) {
         partitionsToMakeFollower.foreach { partition =>
-          stateChangeLogger.trace(s"Skipped the adding-fetcher step of the become-follower state " +
+          stateChangeLogger.info(s"Skipped the adding-fetcher step of the become-follower state " +
```
Let's keep this as TRACE logging here.

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1501,7 +1492,7 @@ class ReplicaManager(val config: KafkaConfig,
         replicaFetcherManager.addFetcherForPartitions(partitionsToMakeFollowerWithLeaderAndOffset)
         partitionsToMakeFollowerWithLeaderAndOffset.foreach { case (partition, initialFetchState) =>
-          stateChangeLogger.trace(s"Started fetcher to new leader as part of become-follower " +
+          stateChangeLogger.info(s"Started fetcher to new leader as part of become-follower " +
```
We already have info level logging in replicaFetcherManager.addFetcherForPartitions(). So we can just remove this logging.

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1515,7 +1506,7 @@ class ReplicaManager(val config: KafkaConfig,
 
     partitionStates.keys.foreach { partition =>
-      stateChangeLogger.trace(s"Completed LeaderAndIsr request correlationId $correlationId from controller $controllerId " +
+      stateChangeLogger.info(s"Completed LeaderAndIsr request correlationId $correlationId from controller $controllerId " +
```
Let's keep this as TRACE logging here.

junrao commented on core/src/main/scala/kafka/controller/ControllerChannelManager.scala:
```diff
@@ -473,10 +471,8 @@ abstract class AbstractControllerBrokerRequestBatch(config: KafkaConfig,
-      stateChangeLog.trace(s"Sending UpdateMetadata request $partitionState to brokers $updateMetadataRequestBrokerSet " +
-        s"for partition $tp")
-    }
+    stateChangeLog.trace(s"Sending UpdateMetadata request to brokers $updateMetadataRequestBrokerSet " +
```
We can probably keep this as trace. While you are at this, could we also add a trace logging to sendStopReplicaRequest()?

stanislavkozlovski replied:
In `sendStopReplicaRequest` we have some debug logs:
```
        debug(s"The stop replica request (delete = true) sent to broker $brokerId is ${stopReplicaWithDelete.mkString(",")}")
```
is this the type of log that we'd prefer the stateChangeLogger to have?

junrao commented on core/src/main/scala/kafka/controller/ControllerChannelManager.scala:
```diff
@@ -266,7 +266,7 @@ class RequestSendThread(val controllerId: Int,
         val response = clientResponse.responseBody
 
-        stateChangeLogger.withControllerEpoch(controllerContext.epoch).trace(s"Received response " +
+        stateChangeLogger.withControllerEpoch(controllerContext.epoch).info(s"Received response " +
```
We can probably keep this as trace.

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -377,17 +377,22 @@ class ReplicaManager(val config: KafkaConfig,
               responseMap.put(topicPartition, Errors.KAFKA_STORAGE_ERROR)
           }
         }
+        stateChangeLogger.info(s"Successfully handled stop replica (delete=${stopReplicaRequest.deletePartitions}) for ${successfulStopReplicaCount} partitions")
```
Since we already have logging in replicaFetcherManager.removeFetcherForPartitions(), we probably don't need any new logging in stopReplicas().

junrao reviewed (commented):
@stanislavkozlovski : Thanks for the updated PR. Made another pass. Since we are seeing many fewer controller related bugs these days, I was thinking that we can reduce the info level logging a bit. Specially, I suggested to leave all request level logging to trace and just keep one partition level logging in handling LeaderAndIsrRequest.

stanislavkozlovski commented on core/src/main/scala/kafka/controller/ControllerChannelManager.scala:
```diff
@@ -550,6 +547,7 @@ abstract class AbstractControllerBrokerRequestBatch(config: KafkaConfig,
         brokerEpoch, deletePartitions, partitions)
     }
 
+    stateChangeLog.trace(s"Sending StopReplica request (version $stopReplicaRequestVersion) to brokers ${controllerContext.liveOrShuttingDownBrokerIds}")
```
I was wondering if the `debug()` logs a couple of lines below are better suited for the stateChangeLogger

junrao replied:
This is covered by the logging below. So, we can just remove this.

stanislavkozlovski commented on core/src/main/scala/kafka/cluster/Partition.scala:
```diff
@@ -497,7 +498,7 @@ class Partition(val topicPartition: TopicPartition,
       val leaderLog = localLogOrException
       val leaderEpochStartOffset = leaderLog.logEndOffset
-      info(s"$topicPartition starts at leader epoch ${partitionState.leaderEpoch} from " +
+      stateChangeLogger.info(s"Leader $topicPartition starts at leader epoch ${partitionState.leaderEpoch} from " +
```
I think this made it clearer and easy to separate from the `makeFollower()` log, even though both have a log line indicating that we're handling a `become-leader` or `become-follower` request

stanislavkozlovski commented:
Thanks for being patient with the review @junrao! You can tell I'm not too familiar with the code.

I believe I've addressed all of your comments. I've got one outstanding question about the per-partition debug logs in ControllerChannelManager#sendStopReplicaRequests()

junrao commented on core/src/main/scala/kafka/controller/ControllerChannelManager.scala:
```diff
@@ -550,6 +547,7 @@ abstract class AbstractControllerBrokerRequestBatch(config: KafkaConfig,
         brokerEpoch, deletePartitions, partitions)
     }
 
+    stateChangeLog.trace(s"Sending StopReplica request (version $stopReplicaRequestVersion) to brokers ${controllerContext.liveOrShuttingDownBrokerIds}")
```
Yes, we could just change the following debug logging to  stateChangeLog.trace. Also, in the logging, we could just log StopReplicaRequestInfo.replica since StopReplicaRequestInfo.deletePartition is already extracted out.

junrao commented on core/src/main/scala/kafka/cluster/Partition.scala:
```diff
@@ -568,6 +569,12 @@ class Partition(val topicPartition: TopicPartition,
       createLogIfNotExists(localBrokerId, partitionState.isNew, isFutureReplica = false, highWatermarkCheckpoints)
 
+      val followerLog = localLogOrException
+      val leaderEpochStartOffset = followerLog.logEndOffset
```
leaderEpochStartOffset => logEndOffset ?

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -377,10 +377,14 @@ class ReplicaManager(val config: KafkaConfig,
         replicaAlterLogDirsManager.removeFetcherForPartitions(partitions)
-        for (topicPartition <- partitions){
+
+        var successfulStopReplicaCount = 0
```
This is no longer needed?

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -377,10 +377,14 @@ class ReplicaManager(val config: KafkaConfig,
-        for (topicPartition <- partitions){
+
+        var successfulStopReplicaCount = 0
+        stateChangeLogger.info(s"Handling stop replica (delete=${stopReplicaRequest.deletePartitions}) for ${partitions.size} partitions (${partitions.mkString(", ")})")
```
We can just log a count at info level.

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1158,12 +1162,10 @@ class ReplicaManager(val config: KafkaConfig,
-          s"correlation id $correlationId from controller ${leaderAndIsrRequest.controllerId} " +
-          s"epoch ${leaderAndIsrRequest.controllerEpoch}")
-      }
+    leaderAndIsrRequest.partitionStates.asScala.foreach { partitionState =>
```
Let's keep this as trace.

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1334,29 +1336,29 @@ class ReplicaManager(val config: KafkaConfig,
       // First stop fetchers for all the partitions
       replicaFetcherManager.removeFetcherForPartitions(partitionStates.keySet.map(_.topicPartition))
+      stateChangeLogger.info(s"Stopped fetchers as part of become-leader LeaderAndIsr request correlationId $correlationId from " +
+        s"controller $controllerId epoch $controllerEpoch starting the become-leader transition for " +
```
We can simplify this a bit to avoid mentioning become-leader twice.

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1467,23 +1472,14 @@ class ReplicaManager(val config: KafkaConfig,
-      }
-
-      if (isShuttingDown.get()) {
+      if (isShuttingDown.get() && traceLoggingEnabled) {
```
Hmm, this seems incorrect. If trace is disabled and shutdown is true, we don't want to add the fetchers in the else clause. So the trace testing needs to be inside `if (isShuttingDown.get()`.

stanislavkozlovski replied:
Oh my, good catch

junrao reviewed (commented):
@stanislavkozlovski : Thanks for the updated PR. A few more comments below.

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -377,7 +377,9 @@ class ReplicaManager(val config: KafkaConfig,
         replicaAlterLogDirsManager.removeFetcherForPartitions(partitions)
-        for (topicPartition <- partitions){
+
+        stateChangeLogger.info(s"Handling stop replica for ${partitions.size} partitions")
```
Could we include StopReplicaRequest.deletePartitions() in the logging?

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1334,29 +1336,29 @@ class ReplicaManager(val config: KafkaConfig,
       // First stop fetchers for all the partitions
       replicaFetcherManager.removeFetcherForPartitions(partitionStates.keySet.map(_.topicPartition))
+      stateChangeLogger.info(s"Stopped fetchers as part of LeaderAndIsr request correlationId $correlationId from " +
+        s"controller $controllerId epoch $controllerEpoch starting the become-leader transition for " +
```
starting the become-leader transition for => as part of the become-leader transition for

junrao commented on core/src/main/scala/kafka/controller/ControllerChannelManager.scala:
```diff
@@ -473,10 +472,8 @@ abstract class AbstractControllerBrokerRequestBatch(config: KafkaConfig,
-      stateChangeLog.trace(s"Sending UpdateMetadata request $partitionState to brokers $updateMetadataRequestBrokerSet " +
-        s"for partition $tp")
-    }
+    stateChangeLog.trace(s"Sending UpdateMetadata request to brokers $updateMetadataRequestBrokerSet " +
```
Should we add `if (stateChangeLog.isTraceEnabled)`?

junrao commented on core/src/main/scala/kafka/controller/ControllerChannelManager.scala:
```diff
@@ -550,19 +547,23 @@ abstract class AbstractControllerBrokerRequestBatch(config: KafkaConfig,
       if (stopReplicaWithDelete.nonEmpty) {
-        debug(s"The stop replica request (delete = true) sent to broker $brokerId is ${stopReplicaWithDelete.mkString(",")}")
+        if (traceEnabled)
+          stateChangeLog.trace(s"The stop replica request (delete = true) sent to broker $brokerId is ${stopReplicaWithDelete.map(_.replica).mkString(",")}")
```
$brokerId is  => $brokerId contains

junrao commented on core/src/main/scala/kafka/controller/ControllerChannelManager.scala:
```diff
@@ -550,19 +547,23 @@ abstract class AbstractControllerBrokerRequestBatch(config: KafkaConfig,
       if (stopReplicaWithoutDelete.nonEmpty) {
-        debug(s"The stop replica request (delete = false) sent to broker $brokerId is ${stopReplicaWithoutDelete.mkString(",")}")
+        if (traceEnabled)
+          stateChangeLog.trace(s"The stop replica request (delete = false) sent to broker $brokerId is ${stopReplicaWithoutDelete.map(_.replica).mkString(",")}")
```
$brokerId is => $brokerId contains

junrao commented on core/src/main/scala/kafka/controller/ControllerChannelManager.scala:
```diff
@@ -575,7 +576,7 @@ abstract class AbstractControllerBrokerRequestBatch(config: KafkaConfig,
       sendLeaderAndIsrRequest(controllerEpoch, stateChangeLog)
       sendUpdateMetadataRequests(controllerEpoch, stateChangeLog)
-      sendStopReplicaRequests(controllerEpoch)
+      sendStopReplicaRequests(controllerEpoch, stateChangeLog)
```
Do we need to pass stateChangeLog in? It seems that we can just reference it directly.

stanislavkozlovski replied:
I was following the convention as used by the other methods. I've removed the passed in log from all methods now and have them instantiate themselves

ijuma replied:
StateChangeLogger has some state. So passing it around is better than duplicating that state in every method. It also avoids allocations.

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1415,14 +1418,16 @@ class ReplicaManager(val config: KafkaConfig,
                             responseMap: mutable.Map[TopicPartition, Errors],
                             highWatermarkCheckpoints: OffsetCheckpoints) : Set[Partition] = {
+    val traceLoggingEnabled = stateChangeLogger.isTraceEnabled
+    stateChangeLogger.info(s"Handling LeaderAndIsr request correlationId $correlationId from controller $controllerId " +
```
To be consistent, could we add the same info logging at the beginning of makeLeaders()?

junrao commented on core/src/main/scala/kafka/controller/ReplicaStateMachine.scala:
```diff
@@ -422,7 +422,7 @@ class ZkReplicaStateMachine(config: KafkaConfig,
   private def logSuccessfulTransition(replicaId: Int, partition: TopicPartition, currState: ReplicaState, targetState: ReplicaState): Unit = {
     stateChangeLogger.withControllerEpoch(controllerContext.epoch)
-      .trace(s"Changed state of replica $replicaId for partition $partition from $currState to $targetState")
+      .info(s"Changed state of replica $replicaId for partition $partition from $currState to $targetState")
```
This is actually not very useful. So we can leave this in trace. What's more useful is to add an info logging for the leaderAndIsr after the removeReplicasFromIsr() call in line 221.

stanislavkozlovski replied:
Went with
```
stateLogger.info(s"Removed replica $replicaId from the ISR of ${updatedLeaderIsrAndControllerEpochs.size} partitions as part of transition to $OfflineReplica")
```

junrao commented on core/src/main/scala/kafka/cluster/Partition.scala:
```diff
@@ -497,7 +498,7 @@ class Partition(val topicPartition: TopicPartition,
       val leaderLog = localLogOrException
       val leaderEpochStartOffset = leaderLog.logEndOffset
-      info(s"$topicPartition starts at leader epoch ${partitionState.leaderEpoch} from " +
+      stateChangeLogger.info(s"Leader $topicPartition starts at leader epoch ${partitionState.leaderEpoch} from " +
```
It would be useful to log isr, addingReplicas and removingReplicas too.

stanislavkozlovski replied:
I was considering adding a toString method to the `AssignmentState` class to make this easier but went without it for now

junrao reviewed (commented):
@stanislavkozlovski : Thanks for the updated PR. Made another pass and left a few more comments.

junrao commented on core/src/main/scala/kafka/controller/ReplicaStateMachine.scala:
```diff
@@ -219,6 +224,7 @@ class ZkReplicaStateMachine(config: KafkaConfig,
           controllerContext.partitionLeadershipInfo.contains(replica.topicPartition)
         }
         val updatedLeaderIsrAndControllerEpochs = removeReplicasFromIsr(replicaId, replicasWithLeadershipInfo.map(_.topicPartition))
+        stateLogger.info(s"Removed replica $replicaId from the ISR of ${updatedLeaderIsrAndControllerEpochs.size} partitions as part of transition to $OfflineReplica")
```
For this one, it's useful to log for each partition.

junrao commented on core/src/main/scala/kafka/controller/PartitionStateMachine.scala:
```diff
@@ -252,13 +252,13 @@ class ZkPartitionStateMachine(config: KafkaConfig,
       case OfflinePartition =>
         validPartitions.foreach { partition =>
-          stateChangeLog.trace(s"Changed partition $partition state from ${partitionState(partition)} to $targetState")
+          stateChangeLog.info(s"Changed partition $partition state from ${partitionState(partition)} to $targetState")
```
Let's keep this trace.

stanislavkozlovski replied:
What's the reasoning for keeping it in trace? Isn't it useful to know when a replica is set to offline?

When do we enter this state - is it only during topic deletion and broker failure? If so, sounds good to keep in TRACE.

I guess in the cases of broker failure it may be useful to log the total number of offline partitions in `KafkaController#onReplicasBecomeOffline`

junrao replied:
Yes, this is called on broker offline. But it only sets the in-memory state. The real work is done when changing ReplicaState to OfflineReplica case, which we have info level logging.

junrao commented on core/src/main/scala/kafka/controller/PartitionStateMachine.scala:
```diff
@@ -252,13 +252,13 @@ class ZkPartitionStateMachine(config: KafkaConfig,
       case NonExistentPartition =>
         validPartitions.foreach { partition =>
-          stateChangeLog.trace(s"Changed partition $partition state from ${partitionState(partition)} to $targetState")
+          stateChangeLog.info(s"Changed partition $partition state from ${partitionState(partition)} to $targetState")
```
Let's keep this trace.

stanislavkozlovski replied:
We enter this state during topic deletion only, right? Makes sense to me

junrao commented on core/src/main/scala/kafka/server/MetadataCache.scala:
```diff
@@ -302,22 +302,26 @@ class MetadataCache(brokerId: Int) extends Logging {
+        val cachedPartitionsCount = newStates.size - deletedPartitions.size
+        stateChangeLogger.info(s"Deleted ${deletedPartitions.size} partitions from metadata cache in response to UpdateMetadata " +
+          s"request sent by controller $controllerId epoch $controllerEpoch with correlation id $correlationId")
+        stateChangeLogger.info(s"Cached leader info for $cachedPartitionsCount partitions in response to UpdateMetadata " +
```
Could we just combine these 2 lines into a single line?

junrao commented on core/src/main/scala/kafka/controller/ControllerChannelManager.scala:
```diff
@@ -572,9 +576,8 @@ abstract class AbstractControllerBrokerRequestBatch(config: KafkaConfig,
-      val stateChangeLog = stateChangeLogger.withControllerEpoch(controllerEpoch)
-      sendLeaderAndIsrRequest(controllerEpoch, stateChangeLog)
-      sendUpdateMetadataRequests(controllerEpoch, stateChangeLog)
+      sendLeaderAndIsrRequest(controllerEpoch)
```
Didn't realize that we have to construct a new stateChangeLog object. So, the existing approach of constructing it once and pass it along seems better.

junrao commented on core/src/main/scala/kafka/controller/ControllerChannelManager.scala:
```diff
@@ -447,14 +448,13 @@ abstract class AbstractControllerBrokerRequestBatch(config: KafkaConfig,
     leaderAndIsrRequestMap.filterKeys(controllerContext.liveOrShuttingDownBrokerIds.contains).foreach {
       case (broker, leaderAndIsrPartitionStates) =>
-        if (stateChangeLog.isTraceEnabled) {
+        if (stateChangeLog.isTraceEnabled)
```
Could we add an info level logging with the become-leader/become-follower counts for each broker?

junrao commented on core/src/main/scala/kafka/controller/ControllerChannelManager.scala:
```diff
@@ -472,11 +472,11 @@ abstract class AbstractControllerBrokerRequestBatch(config: KafkaConfig,
-    }
+  private def sendUpdateMetadataRequests(controllerEpoch: Int): Unit = {
+    val stateChangeLog = stateChangeLogger.withControllerEpoch(controllerEpoch)
+    if (stateChangeLog.isTraceEnabled)
```
Since this just has the count, we can make this info level.

junrao commented on core/src/main/scala/kafka/controller/ControllerChannelManager.scala:
```diff
@@ -550,19 +551,22 @@ abstract class AbstractControllerBrokerRequestBatch(config: KafkaConfig,
       if (stopReplicaWithoutDelete.nonEmpty) {
-        debug(s"The stop replica request (delete = false) sent to broker $brokerId is ${stopReplicaWithoutDelete.mkString(",")}")
+        if (traceEnabled)
+          stateChangeLog.trace(s"The stop replica request (delete = false) sent to broker $brokerId contains ${stopReplicaWithoutDelete.map(_.replica).mkString(",")}")
```
To be consistent with the above changes, let's add an info level logging per broker with just the partition count.

junrao reviewed (commented):
@stanislavkozlovski : Thanks for the new updates. A few more minor comments below.

junrao commented on core/src/main/scala/kafka/controller/ReplicaStateMachine.scala:
```diff
@@ -220,6 +225,7 @@ class ZkReplicaStateMachine(config: KafkaConfig,
         }
         val updatedLeaderIsrAndControllerEpochs = removeReplicasFromIsr(replicaId, replicasWithLeadershipInfo.map(_.topicPartition))
         updatedLeaderIsrAndControllerEpochs.foreach { case (partition, leaderIsrAndControllerEpoch) =>
+          stateLogger.info(s"Removed replica $replicaId from the ISR of partition $partition $leaderIsrAndControllerEpoch as part of transition to $OfflineReplica")
```
To make it clearer, "Partition xxx state changed to $leaderIsrAndControllerEpoch after removing replica ...".

junrao commented on core/src/main/scala/kafka/server/MetadataCache.scala:
```diff
@@ -302,22 +302,25 @@ class MetadataCache(brokerId: Int) extends Logging {
+          val tp = new TopicPartition(state.topicName, state.partitionIndex)
+          if (state.leader == LeaderAndIsr.LeaderDuringDelete) {
             removePartitionInfo(partitionStates, tp.topic, tp.partition)
-            stateChangeLogger.trace(s"Deleted partition $tp from metadata cache in response to UpdateMetadata " +
```
It's probably useful to leave the per partition logging in trace, under the` if (isTraceEnabled)` flag.

junrao commented on core/src/main/scala/kafka/server/MetadataCache.scala:
```diff
@@ -302,22 +302,25 @@ class MetadataCache(brokerId: Int) extends Logging {
             deletedPartitions += tp
           } else {
-            addOrUpdatePartitionInfo(partitionStates, tp.topic, tp.partition, info)
-            stateChangeLogger.trace(s"Cached leader info $info for partition $tp in response to " +
```
It's probably useful to leave the per partition logging in trace.

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1334,29 +1336,31 @@ class ReplicaManager(val config: KafkaConfig,
                           correlationId: Int,
                           responseMap: mutable.Map[TopicPartition, Errors],
                           highWatermarkCheckpoints: OffsetCheckpoints): Set[Partition] = {
+    val traceEnabled = stateChangeLogger.isTraceEnabled
```
Could we change the following logging in line 1245 to info?

`              stateChangeLogger.debug(s"Ignoring LeaderAndIsr request from "`

stanislavkozlovski replied:
Sounds good, done in https://github.com/apache/kafka/pull/8320/commits/a3469120ffb00fb04d5c4878d991ba685e3bd066

junrao commented on core/src/main/scala/kafka/server/MetadataCache.scala:
```diff
@@ -302,22 +302,25 @@ class MetadataCache(brokerId: Int) extends Logging {
         }
-        metadataSnapshot = MetadataSnapshot(partitionStates, controllerId, aliveBrokers, aliveNodes)
+        val cachedPartitionsCount = newStates.size - deletedPartitions.size
+        stateChangeLogger.info(s"Deleted ${deletedPartitions.size} partitions from metadata cache and " +
```
Deleted 0 partitions from metadata cache and cached leader info for 2 partitions in response => Deleted 0 partitions and add 2 partitions in metadata cache in response

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1415,14 +1420,16 @@ class ReplicaManager(val config: KafkaConfig,
                             responseMap: mutable.Map[TopicPartition, Errors],
                             highWatermarkCheckpoints: OffsetCheckpoints) : Set[Partition] = {
+    val traceLoggingEnabled = stateChangeLogger.isTraceEnabled
+    stateChangeLogger.info(s"Handling LeaderAndIsr request correlationId $correlationId from controller $controllerId " +
```
Let's just add an info level logging with partition count in the caller becomeLeaderOrFollower() instead of at the beginning of makeLeader() and makeFollowers(). This is because sometimes we ignore partitions in becomeLeaderOrFollower().

junrao reviewed (commented):
@stanislavkozlovski : Thanks for the latest PR. Just a few more minor comments below.

stanislavkozlovski commented on core/src/main/scala/kafka/server/MetadataCache.scala:
```diff
@@ -302,22 +302,32 @@ class MetadataCache(brokerId: Int) extends Logging {
             removePartitionInfo(partitionStates, tp.topic, tp.partition)
-            stateChangeLogger.trace(s"Deleted partition $tp from metadata cache in response to UpdateMetadata " +
-              s"request sent by controller $controllerId epoch $controllerEpoch with correlation id $correlationId")
+            if (traceEnabled)
```
Should we add a comment that these trace logs are expensive?

junrao replied:
Yes, we can add a comment.

stanislavkozlovski commented:
@junrao apologies for having to squash all commits - I needed to rebase with master. There was a simple conflict in `ReplicaManager#1381` - the parameters for the method in the  `if (partition.makeLeader(partitionState, highWatermarkCheckpoints))` line

I addressed your last comments in these three commits, before squashing them - https://github.com/stanislavkozlovski/kafka/commit/11f0ed59c1f3f4d9998cd208e82db6f35be6828c, https://github.com/stanislavkozlovski/kafka/commit/09e9cf69a6d3e0cbd1d0beb5549027cbb0f103c4, https://github.com/stanislavkozlovski/kafka/commit/ccd1e38bf17f9a789d30506298c006995d1de494

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1255,15 +1257,19 @@ class ReplicaManager(val config: KafkaConfig,
         val highWatermarkCheckpoints = new LazyOffsetCheckpoints(this.highWatermarkCheckpoints)
-        val partitionsBecomeLeader = if (partitionsTobeLeader.nonEmpty)
+        val partitionsBecomeLeader = if (partitionsTobeLeader.nonEmpty) {
+          stateChangeLogger.info(s"Handling LeaderAndIsr request correlationId $correlationId from controller $controllerId " +
```
I was thinking of logging the partition count at the beginning of becomeLeaderOrFollower(). Otherwise, it's kind of weird to see the request logging after the logging for "Ignoring".

junrao commented on core/src/main/scala/kafka/server/MetadataCache.scala:
```diff
@@ -321,22 +321,32 @@ class MetadataCache(brokerId: Int) extends Logging {
         }
-        metadataSnapshot = MetadataSnapshot(partitionStates, controllerId, aliveBrokers, aliveNodes)
+        val cachedPartitionsCount = newStates.size - deletedPartitions.size
+        stateChangeLogger.info(s"Deleted ${deletedPartitions.size} partitions and " +
```
Could we put "add $cachedPartitionsCount partitions" before "Deleted ${deletedPartitions.size} partitions" ?

junrao reviewed (commented):
@stanislavkozlovski : Thanks for the new update. Just a few more comments below.

stanislavkozlovski committed:
Move logging in becomeLeaderOrFollower

stanislavkozlovski committed:
Tweak logging in MetadataCache

junrao commented:
ok to test

junrao reviewed (approved):
@stanislavkozlovski : Thanks for the latest PR. LGTM. Wait for tests to pass.

junrao commented:
The default state-change log looks like the following.

```
[2020-03-26 11:38:33,075] INFO [Controller id=0 epoch=1] Sending LeaderAndIsr request to broker 1 with 0 become-leader and 1 become-follower partitions (state.change.l
ogger)
[2020-03-26 11:38:33,075] INFO [Controller id=0 epoch=1] Sending LeaderAndIsr request to broker 0 with 1 become-leader and 0 become-follower partitions (state.change.l
ogger)
[2020-03-26 11:38:33,075] INFO [Controller id=0 epoch=1] Sending UpdateMetadata request to brokers Set(0, 1) for 1 partitions (state.change.logger)
[2020-03-26 11:38:33,076] INFO [Broker id=0] Handling LeaderAndIsr request correlationId 4 from controller 0 for 1 partitions (state.change.logger)
[2020-03-26 11:38:33,078] INFO [Broker id=0] Stopped fetchers as part of LeaderAndIsr request correlationId 4 from controller 0 epoch 1 as part of the become-leader tr
ansition for 1 partitions (state.change.logger)
[2020-03-26 11:38:33,078] INFO [Broker id=0] Leader test-1 starts at leader epoch 1 from offset 0 with high watermark 0 ISR 0 addingReplicas  removingReplicas .Previous leader epoch was 0. (state.change.logger)
[2020-03-26 11:38:33,079] INFO [Controller id=0 epoch=1] Sending UpdateMetadata request to brokers Set() for 0 partitions (state.change.logger)
[2020-03-26 11:38:33,080] INFO [Controller id=0 epoch=1] Sending a stop replica request (delete = false) for 1 replicas to broker 1 (state.change.logger)
[2020-03-26 11:38:33,088] INFO [Broker id=0] Add 1 partitions and deleted 0 partitions from metadata cache in response to UpdateMetadata request sent by controller 0 epoch 1 with correlation id 5 (state.change.logger)
[2020-03-26 11:38:33,095] INFO [Controller id=0 epoch=1] Partition test-0 state changed to (Leader:0,ISR:0,LeaderEpoch:1,ControllerEpoch:1) after removing replica 1 from the ISR as part of transition to OfflineReplica (state.change.logger)
[2020-03-26 11:38:33,096] INFO [Controller id=0 epoch=1] Sending LeaderAndIsr request to broker 0 with 1 become-leader and 0 become-follower partitions (state.change.logger)
[2020-03-26 11:38:33,097] INFO [Controller id=0 epoch=1] Sending UpdateMetadata request to brokers Set(0, 1) for 1 partitions (state.change.logger)
[2020-03-26 11:38:33,097] INFO [Controller id=0 epoch=1] Sending a stop replica request (delete = false) for 1 replicas to broker 1 (state.change.logger)
[2020-03-26 11:38:33,097] INFO [Broker id=0] Handling LeaderAndIsr request correlationId 6 from controller 0 for 1 partitions (state.change.logger)
[2020-03-26 11:38:33,097] INFO [Broker id=0] Stopped fetchers as part of LeaderAndIsr request correlationId 6 from controller 0 epoch 1 as part of the become-leader transition for 1 partitions (state.change.logger)
[2020-03-26 11:38:33,098] INFO [Broker id=0] Leader test-0 starts at leader epoch 1 from offset 0 with high watermark 0 ISR 0 addingReplicas  removingReplicas .Previous leader epoch was 0. (state.change.logger)
[2020-03-26 11:38:33,099] INFO [Broker id=0] Skipped the become-leader state change after marking its partition as leader with correlation id 6 from controller 0 epoch 1 for partition test-0 (last update controller epoch 1) since it is already the leader for the partition. (state.change.logger)
[2020-03-26 11:38:33,100] INFO [Broker id=0] Add 1 partitions and deleted 0 partitions from metadata cache in response to UpdateMetadata request sent by controller 0 epoch 1 with correlation id 7 (state.change.logger)

```

junrao commented:
Only one following failure. Seems unrelated.

```
12:29:16 org.apache.kafka.streams.processor.internals.RecordCollectorTest > shouldNotThrowStreamsExceptionOnSubsequentCallIfASendFailsWithContinueExceptionHandler FAILED
12:29:16     java.lang.AssertionError
12:29:16         at org.junit.Assert.fail(Assert.java:87)
12:29:16         at org.junit.Assert.assertTrue(Assert.java:42)
12:29:16         at org.junit.Assert.assertTrue(Assert.java:53)
12:29:16         at org.apache.kafka.streams.processor.internals.RecordCollectorTest.shouldNotThrowStreamsExceptionOnSubsequentCallIfASendFailsWithContinueExceptionHandler(RecordCollectorTest.java:519)
```
